dotnet/Razorvine.Pyrolite/Benchmarks/Benchmarks.csproj

```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
```

Comment:
Should be `2.1`.  `2.0` is EOL'd.

Reply:
I wanted to be consistent with other projects in this solution and repository (they all target .NET Core 2.0)

https://github.com/irmen/Pyrolite/blob/060bc3c9069cd31560b6da4f67280736fb2cdf63/dotnet/Razorvine.Pyrolite/Tests/Tests.csproj#L3

Reply:
Maybe it would make sense to log a bug to move them all to `2.1`. It doesn't really make sense to be testing on a dead version of the framework.

Reply:
@irmen would you accept a PR with update from .NET Core 2.0 to 2.1?

Reply:
Where can I read about 2.0 being EOL? 
Also, from what I can find, 2.1 will require Mono 6 (which isn't even released yet) and I require  Pyrolite to be compatible with mono 5 for the time being.

Reply:
## LangVersion

C# is compiled to IL (byte code) and can be executed by a .NET Runtime: .NET Core, Mono or Full .NET Framework.

The `LangVersion` setting set's the version of C# compiler. Most of the new features are using existing IL features, so updating the `LangVersion` does not require the users of a library to update the Runtime they are using. This is going to change with C# 8.0 which introduces features that require new Runtime. But setting LangVersion to 7.3 means use latest C# compiler that does not have any runtime-specific features. In my opinion it's the best choice for libraries as of today 👍 

## .NET Standard

Before .NET Core was born, the .NET libraries were targeting a particular version of Full .NET Framework: `net462`, `net47` etc.  Mono was able to run such libraries without explicit compilation target for Mono (Mono was doing all the heavy lifting behind the scenes).

But when .NET Core was born it became clear that all the 3 .NET Runtimes offer a very simlar set of public APIs. 
This set was called .NET Standard.
A library that targets a particular version of .NET Standard can run on any .NET Runtime that supports it. ([full matrix](https://docs.microsoft.com/en-us/dotnet/standard/net-standard))
When the library is compiled for .NET Standard it does not reference any runtime, but a set of reference libraries. When a .NET Standard code is executed, the .NET Runtime uses the right implementation of reference assemblies.

From a developer perspective I would use an analogy of abstractions, their implementations and dependency containers. When you write a .NET Standard code you reference and compile against "abstraction". When you run a .NET Standard code the .NET Runtime acst like a dependency container that injects the proper implementations.

The Pyrloite library targets .NET Standard 2.0 which in my opinion is the best choice for libraries as of today 👍 

https://github.com/irmen/Pyrolite/blob/56b8163f38b2e439385b909d10af1cc824e19ec6/dotnet/Razorvine.Pyrolite/Pyrolite/Pyrolite.csproj#L3

Which means that it can be used by .NET applications targeting:
- .NET Framework 4.6.1+
- Mono 5.4+
- .NET Core 2.0+

## Applications

The applications can't target .NET Standard, but particular runtime(s). They need to specify the execution runtime. Using my previous analogy you can't create an instance of interface, but you can create an instance of class that implements the interface.

@irmen now back to your question what @eerhardt offered was to update the Test and Benchmarks console apps from .NET Core 2.0 to 2.1. Not to update Pyrloite library from .NET Standard 2.0 to .NET Standard 2.1.

Such change would not affect the consumers of the Pyrloite library, but the contributors who would need to install .NET Core 2.1 SDK to run the tests and benchmarks. .NET Core 2.0 is EOL, but .NET Standard 2.0 is not and will most probably never be. The future runtimes are going to extend .NET Standard with new APIs and create new versions of .NET Standard, but they still will be supersets of it (not subsets).

Reply:
Thanks a lot for the extensive explanation to clear that up for me, I got a bit confused there!

Reply:
> Where can I read about 2.0 being EOL?

https://devblogs.microsoft.com/dotnet/net-core-2-0-will-reach-end-of-life-on-september-1-2018/

(NOTE the difference between `netstandard2.0` which is very alive and well, and `netcoreapp2.0` which is now replaced by `netcoreapp2.1`.)